Fix handling for json-file io.UnexpectedEOF

cpuguy83 opened the pull request:
Before this, we essentially end up doubling any buffered data from the
json decoder by storing it in both `d.dec` and `d.rdr`.

Thanks @tobiasstadler for pointing this error out.

Fixes #41820

cpuguy83 committed:
Fix handling for json-file io.UnexpectedEOF

When the multireader hits EOF, we will always get EOF from it, so we
cannot store the multrireader fro later error handling, only for the
decoder.

Thanks @tobiasstadler for pointing this error out.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

thaJeztah commented:
s/io.UnepxectedEOF/io.UnexpectedEOF/ ☺️

thaJeztah commented:
@tonistiigi @tiborvass ptal

thaJeztah commented:
@cpuguy83 regression was introduced in https://github.com/moby/moby/pull/40796, correct?

cpuguy83 commented:
@thaJeztah That's correct.
Test is passes before that commit (but requires some modifications due to refactoring in said PR), and fails on that PR and master.

tonistiigi commented on daemon/logger/jsonfilelog/read.go:
```diff
@@ -105,8 +110,7 @@ func (d *decoder) Decode() (msg *logger.Message, err error) {
 		if err == io.ErrUnexpectedEOF {
-			d.rdr = io.MultiReader(d.dec.Buffered(), d.rdr)
-			d.dec = json.NewDecoder(d.rdr)
+			d.dec = json.NewDecoder(io.MultiReader(d.dec.Buffered(), d.rdr))
```
iiuc the issue in here is not the resetting `d.rdr` or not (what looks more correct to me in the old version) but that we use incompatible `Reader` where after it returns `io.EOF` we can read again and it may return more data then. This does not follow the `io.Reader` interface that stdlib defines and therefore the utility functions behavior gets wrong. More specifically `io.MultiReader` caches `io.EOF` https://github.com/golang/go/blob/414fa8c35e7c2f65e2c767d6db2f25791e53b5c1/src/io/multi.go#L41 . So in the previous code if the `MultiReader` has already returned `EOF` once and now is added as a second reader to a new `MultiReader` it never gets called again and new `MultiReader` returns `EOF` without making any new `Read` calls to the original `d.rdr`.

cpuguy83 replied:
Added a 2nd commit per your suggestion to replace io.MultiReader with a custom implementation that does not cache EOF's so that we can keep the full reader (with the buffered data) in `d.rdr`.

cpuguy83 committed:
jsonfile: more defensive reader implementation

Tonis mentioned that we can run into issues if there is more error
handling added here. This adds a custom reader implementation which is
like io.MultiReader except it does not cache EOF's.
What got us into trouble in the first place is `io.MultiReader` will
always return EOF once it has received an EOF, however the error
handling that we are going for is to recover from an EOF because the
underlying file is a file which can have more data added to it after
EOF.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

tonistiigi reviewed (approved):
LGTM

`SyntaxError` handler seems to require `combinedReader` as well but that can be done separately as not directly related. Although I wonder if it might be broken because it could not have been made to work before because of the multireader issue.

cpuguy83 commented:
Yeah there's no test for syntax reader and I think it never actually fixed anything.

thaJeztah commented on daemon/logger/jsonfilelog/read.go:
```diff
@@ -105,14 +110,54 @@ func (d *decoder) Decode() (msg *logger.Message, err error) {
 	return msg, err
 }
 
+func comineReaders(pre, rdr io.Reader) io.Reader {
```
```suggestion
func combineReaders(pre, rdr io.Reader) io.Reader {
```

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
CI was green, and only typo was fixed in last push

cpuguy83 commented:
> === FAIL: amd64.integration.service TestInspect (11.75s)
[2021-03-18T19:09:12.879Z]     inspect_test.go:39: timeout hit after 10s: waiting for tasks to enter run state. task failed with error: task: non-zero exit (1)


This flaky inspect test :(

anodyne-canvas commented:
When will this be included in a release?

thaJeztah commented:
It's backported to the 20.10 release branch through https://github.com/moby/moby/pull/42174, and will be included in the 20.10.6 patch release

anodyne-canvas commented:
Thanks for following up, do you know when 20.10.6 will be released?

aleon1220 commented:
It seems to be fixed in docker engine 20.10.6 https://docs.docker.com/engine/release-notes/#20106

BenasPaulikas commented:
almost 2am here. updating docker with this patch. Praying for this fix to solve my issues :)

thaJeztah commented:
@BenasPaulikas docker 20.10.6 and 20.10.7 have already shipped with this patch, so updating to the latest patch release should bring this patch.

oren-nonamesecurity commented:
Same here, docker 20.10.10 flooding:
```
![image](https://user-images.githubusercontent.com/62993794/198866772-55eef0d0-a134-408a-8dd8-4a95c5a44f7f.png)

```

oren-nonamesecurity commented:
Same here, docker 20.10.10
![Screenshot from 2022-10-30 09-10-58](https://user-images.githubusercontent.com/62993794/198866819-64334c0e-43de-49ba-8816-2c31de5e3d75.png)

BenasPaulikas commented:
I might be wrong here.

But there was package that was using json to get docker output or something. They changed to use streams (or something simmilar) and issue was gone. + It saved so much CPU power

thaJeztah commented:
> I'm still seeing this issue on 20.10.17, and so are many people on [#41820 (comment)](https://github.com/moby/moby/issues/41820#issuecomment-1142518992)
> Any chance this could be reviewed/reopened?

This PR was to address an issue where docker tried to read the JSON, got an EOF, and reused the reader (which would continue to return an EOF error). Note that the EOF error can be an actual issue with a corrupt file, in which case the `EOF` error would be legit.

xtemp3r commented:
> I'm still seeing this issue on 20.10.17, and so are many people on [#41820 (comment)](https://github.com/moby/moby/issues/41820#issuecomment-1142518992)

I can confirm I have this problem again on multiple systems running docker. I would also suggest that this issue is reopened.

I'm also running version: 20.10.17

fredsig commented:
Same issue here with docker 20.10.18. Getting a couple of warnings every second (but it is not affecting my ability to process container logs) This triggered by using Loki's promtail agent to collect docker logs using the API (aka, "docker logs') via the service discovery plugin for docker:

```
Apr  2 16:23:52 ip-192-168-151-3 dockerd[323961]: time="2024-04-02T16:23:52.260711951Z" level=warning msg="got error while decoding json" error="unexpected EOF" retries=0
Apr  2 16:23:52 ip-192-168-151-3 dockerd[323961]: time="2024-04-02T16:23:52.587555447Z" level=warning msg="got error while decoding json" error="unexpected EOF" retries=0
Apr  2 16:23:52 ip-192-168-151-3 dockerd[323961]: time="2024-04-02T16:23:52.648610496Z" level=warning msg="got error while decoding json" error="unexpected EOF" retries=0
```
